core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -174,60 +243,89 @@ public boolean hasListeners() {
   }
 
   private final Set<EventListenerWrapper> registeredListeners = new CopyOnWriteArraySet<EventListenerWrapper>();
-  private final ExecutorService orderedDelivery;
-  private final ExecutorService unorderedDelivery;
-
-  private static final class EventListenerWrapper {
-    final CacheEventListener<?, ?> listener;
-    final CacheEventListenerConfiguration config;
-    
-    EventListenerWrapper(CacheEventListener<?, ?> listener, final EventFiring firing, final EventOrdering ordering, 
-        final EnumSet<EventType> forEvents) {
-      this.listener = listener;
-      this.config = new CacheEventListenerConfiguration() {
-        
-        @Override
-        public Class<CacheEventListenerProvider> getServiceType() {
-          return CacheEventListenerProvider.class;
-        }
-        
-        @Override
-        public EventOrdering orderingMode() {
-          return ordering;
-        }
-        
-        @Override
-        public EventFiring firingMode() {
-          return firing;
-        }
-        
-        @Override
-        public EnumSet<EventType> fireOn() {
-          return forEvents;
-        }
-      };
+  private final Set<CacheEventListener> orderedListenersSet = new CopyOnWriteArraySet<CacheEventListener>();
+  private final Set<EventListenerWrapper> syncListenersSet = new CopyOnWriteArraySet<EventListenerWrapper>();
+  private final Set<EventListenerWrapper> aSyncListenersSet = new CopyOnWriteArraySet<EventListenerWrapper>();
+
+  private boolean addOrderedListenerToSet(CacheEventListener<? super K, ? super V> listener, EventOrdering ordering) {
+    eventDispatcher = eventDispatchers.get(1);
+    return ordering != EventOrdering.ORDERED || orderedListenersSet.add(listener);
+  }
+  
+  private boolean removeOrderedListenerFromSet(CacheEventListener<? super K, ? super V> listener) {
+    orderedListenersSet.remove(listener);
+    if (orderedListenersSet.isEmpty()) {
```

Comment:
Given this is the condition for interacting with the ordered queues, what happens when you have elements in the queues and this becomes true?